File: witnessgame.cpp
#include <vector>
#include <map>

#include "object.h"
#include "miscsymbols.h"
#include "blockgroup.h"
#include "util.h"
#include "grid.h"
#include "solver.h"
#include "randgrid.h"

#include "witnessgame.h"

WitnessGame::WitnessGame(Grid g) {
    grid = g;
    startpts.clear();
    for (auto i : grid.starts) startpts.push_back(i);
    std::sort(startpts.begin(), startpts.end());
    origin = startpts[0];
    pos = origin;
}

void WitnessGame::reset(Grid g) {
    grid = g;
    startpts.clear();
    for (auto i : grid.starts) startpts.push_back(i);
    std::sort(startpts.begin(), startpts.end());
    origin = startpts[0];
    pos = origin;
}

void WitnessGame::setStart(int x) {
    if (startpts.size() <= 0) return;
    while (x < 0) x += startpts.size();
    origin = startpts[x % startpts.size()];
}

void WitnessGame::processInput(char c) { // Char to dz direction
    int dir = -1;
    for (int i = 0; i < 4; i++) {
        if (c == dz[i]) dir = i;
    }
    if (dir < 0) {
        std::cout << "INVALID DIRECTION\n";
        return;
    }
    std::pair<int, int> between = {pos.first + dx[dir], pos.second + dy[dir]};
    std::pair<int, int> hit = {pos.first + 2 * dx[dir], pos.second + 2 * dy[dir]};
    if (!grid.inside(between) || !grid.inside(hit)) {
        std::cout << "ERROR OUT OF BOUNDS\n";
        return;
    }
    if (!grid.board[between.first][between.second]->isPath || !grid.board[hit.first][hit.second]->isPath) {
        std::cout << "ERROR NOT VALID MOVE\n";
        return;
    }
    
    if (parent.find(pos) != parent.end() && parent.at(pos) == hit) {
        grid.board[pos.first][pos.second]->isPathOccupied = false;
        grid.board[between.first][between.second]->isPathOccupied = false;
        parent.erase(parent.find(pos));
        pos = hit;
        return;
    }
    std::cout << "[" << hit.first << " " << hit.second << "/" << pos.first << " " << pos.second << "]\n";
    parent.insert({hit, pos});
    grid.drawLine(pos, hit);
    pos = hit;
}

bool WitnessGame::reachedEnd() {
    return grid.ends.find(pos) != grid.ends.end();
}

void WitnessGame::clear() {
    startpts.clear();
    for (auto i : grid.starts) startpts.push_back(i);
    std::sort(startpts.begin(), startpts.end());
    origin = startpts[0];
    pos = origin;
    for (int i = 0; i < grid.board.size(); i++) {
        for (int j = 0; j < grid.board[i].size(); j++) grid.board[i][j]->isPathOccupied = false;
    }
}

File: PYTHON/witnessgame.py
from object import Object, Color
from miscsymbols import Endpoint, Dot, Blob, Star, Triangle, Cancel
from blockgroup import BlockGroup
from grid import Grid
import util

class WitnessGame:
	grid = 0
	parent = dict()
	pos = (0, 0)
	origin = (0, 0)
	startingPoints = []

	dy = [1, 0, -1, 0]
	dx = [0, 1, 0, -1]
	dz = ['R', 'D', 'L', 'U']
	
	def __init__(self, g):
		self.grid = g
		self.parent = dict()

		self.startingPoints = []
		for p in g.starts:
			self.startingPoints.append(p)
		self.startingPoints.sort()
		self.origin = self.startingPoints[0]
		self.pos = self.origin
	
	def reset(self, g):
		self.grid = g
		self.parent = dict()

		self.startingPoints = []
		for p in g.starts:
			self.startingPoints.append(p)
		self.startingPoints.sort()
		self.origin = self.startingPoints[0]
		self.pos = self.origin
	
	def setStart(self, i):
		if len(self.startingPoints) <= 0:
			return 
		while i < 0:
			i = i + len(self.startingPoints)
		self.origin = self.startingPoints[i % len(self.startingPoints)]
	
	def processInput(self, c):
		dir = -1
		for i in range(4):
			if (c == self.dz[i]):
				dir = i
		
		if dir < 0:
			print("INVALID DIRECTION")
			return
		
		between = (self.pos[0] + self.dx[dir], self.pos[1] + self.dy[dir])
		hit = (self.pos[0] + 2 * self.dx[dir], self.pos[1] + 2 * self.dy[dir])
		if (not self.grid.inside(between)) or (not self.grid.inside(hit)):
			print("ERROR OUT OF BOUNDS")
			return
		
		if (not self.grid.get(between).isPath) or (not self.grid.get(hit).isPath):
			print("ERROR INVALID MOVE")
			return
		
		if self.pos in self.parent:
			if self.parent[self.pos] == hit:
				self.grid.board[self.pos[0]][self.pos[1]].isPathOccupied = False
				self.grid.board[between[0]][between[1]].isPathOccupied = False
				self.parent.pop(self.pos)
				self.pos = hit
				return
		
		print(hit, "/", self.pos)
		self.parent[hit] = self.pos
		self.grid.drawLine(self.pos, hit)
		self.pos = hit

	def reachedEnd(self):
		return self.pos in self.grid.ends

	def clear(self):
		for i in range(len(self.grid.board)):
			for j in range(len(self.grid.board[i])):
				self.grid.board[i][j].isPathOccupied = False
		
		self.grid.violations.clear()
		self.pos = self.origin

